### PR TITLE
[corlib] Properly implement Environment.Is64BitOperatingSystem

### PR DESCRIPTION
**Pull request by akoeplinger**
Before we only checked if IntPtr.Size == 8, but this returns false for a 32bit Mono running on a 64bit OS.
Added an icall that gets the correct value from the runtime instead.

Fixes https://bugzilla.xamarin.com/show_bug.cgi?id=37414

### CODE REVIEW

File: mono/metadata/icall.c [hunk 1]
```diff
@@ -5888,6 +5892,28 @@ ves_icall_System_Environment_get_NewLine (void)
...
 
+ICALL_EXPORT MonoBoolean
+ves_icall_System_Environment_GetIs64BitOperatingSystem (void)
+{
+#if SIZEOF_VOID_P == 8
+	return TRUE;
+#else
+#ifdef HOST_WIN32
+	MonoBoolean isWow64Process = FALSE;
+	if (IsWow64Process (GetCurrentProcess (), &isWow64Process) {
```
Comment (tritao) [inline mono/metadata/icall.c]: I think this is wrong. `MonoBoolean` is a `mono_byte`, which in turn is a `uint8_t`, while `IsWow64Process` parameter is a `BOOL`, aka `int`.

Comment (akoeplinger) [inline mono/metadata/icall.c]: I replaced it with `gboolean` to be consistent with the other usages in this file. I tried it on my windows box and it works.

File: mcs/class/corlib/System/Environment.cs [hunk 2]
```diff
@@ -900,8 +900,12 @@ public static void FailFast (string message, Exception exception)
 			throw new NotImplementedException ();
 		}
 
+		[MethodImplAttribute (MethodImplOptions.InternalCall)]
+		extern static bool GetIs64BitOperatingSystem ();
+		static readonly Lazy<bool> is64BitOperatingSystem = new Lazy<bool> (GetIs64BitOperatingSystem);
```
Comment (kumpera) [inline mcs/class/corlib/System/Environment.cs]: No need for caching here. This adds a dependency on Lazy when we don't need it.

File: mono/metadata/icall.c [hunk 3]
```diff
@@ -5888,6 +5892,28 @@ ves_icall_System_Environment_get_NewLine (void)
...
+#ifdef HOST_WIN32
+	gboolean isWow64Process = FALSE;
+	if (IsWow64Process (GetCurrentProcess (), &isWow64Process)) {
+		return (MonoBoolean)isWow64Process;
+	}
+#elif defined(HAVE_SYS_UTSNAME_H)
+	struct utsname name;
+
+	if (uname (&name) >= 0) {
+		return strcmp (name.machine, "x86_64") == 0;
```
Comment (kumpera) [inline mono/metadata/icall.c]: This is wrong as it won't work on any other 64bits platforms like arm64.

Comment (akoeplinger) [inline mono/metadata/icall.c]: I added `aarch64` for arm64.

**Comment by tritao** [comment]
Do we need to increase the corlib version here?

**Comment by akoeplinger** [comment]
@tritao no, adding/removing icalls can be done without a corlib version bump.

File: mono/metadata/icall.c [hunk 4]
```diff
@@ -5888,6 +5892,28 @@ ves_icall_System_Environment_get_NewLine (void)
...
+#ifdef HOST_WIN32
+	gboolean isWow64Process = FALSE;
+	if (IsWow64Process (GetCurrentProcess (), &isWow64Process)) {
+		return (MonoBoolean)isWow64Process;
+	}
+#elif defined(HAVE_SYS_UTSNAME_H)
+	struct utsname name;
+
+	if (uname (&name) >= 0) {
+		return strcmp (name.machine, "x86_64") == 0 || strcmp (name.machine, "aarch64") == 0;
```
Comment (alexrp) [inline mono/metadata/icall.c]: PPC64 is missing here.

Comment (akoeplinger) [inline mono/metadata/icall.c]: Updated.

**Comment by tritao** [comment]
LGTM.

**Pull request merged**